networkit/cpp/centrality/DynApproxBetweenness.cpp

```diff
@@ -1,178 +1,268 @@
+            vdapproxs[c] += vdSuccessors[c];
+        }
+        if (vdapproxs[c] > vd) {
+            vd = vdapproxs[c];
+        }
+        sccDAG.forInEdgesOf(c, [&](node c, node c_pred) {
+            if (vdSuccessors[c_pred] < vdapproxs[c])
+                vdSuccessors[c_pred] = vdapproxs[c];
+        });
+    }
+    return vd;
+}
 
 count DynApproxBetweenness::getNumberOfSamples() {
     return r;
 }
 
-
 void DynApproxBetweenness::run() {
-    if (G.isDirected()) {
-        throw std::runtime_error("Invalid argument: G must be undirected.");
-    }
     scoreData.clear();
     scoreData.resize(G.upperNodeIdBound());
     u.clear();
     v.clear();
     sampledPaths.clear();
-
-    Diameter diam(G, DiameterAlgo::estimatedPedantic);
-    diam.run();
-    edgeweight vd = diam.getDiameter().first;
-
-    INFO("estimated diameter: ", vd);
-    r = ceil((universalConstant / (epsilon * epsilon)) * (floor(log2(vd - 2)) + 1 - log(delta)));
-    INFO("taking ", r, " path samples");
+    count vd;
+    if (G.isDirected()) {
+        vd = computeVDdirected();
+    } else {
+        Diameter diam(G, DiameterAlgo::estimatedPedantic);
+        diam.run();
+        vd = diam.getDiameter().first;
+    }
+    r = ceil((universalConstant / (epsilon * epsilon))
+             * (floor(log2(vd - 2.0)) + 1.0 + log(1.0 / delta)));
     sssp.clear();
     sssp.resize(r);
-    u.resize(r);
-    v.resize(r);
-    sampledPaths.resize(r);
-
-    for (count i = 0; i < r; i++) {
-        DEBUG("sample ", i);
-        // sample random node pair
-        u[i] = GraphTools::randomNode(G);
-        do {
-            v[i] = GraphTools::randomNode(G);
-        } while (v[i] == u[i]);
-        if (G.isWeighted()) {
-            sssp[i] = std::make_unique<DynDijkstra>(G, u[i], storePreds);
-        } else {
-            sssp[i] = std::make_unique<DynBFS>(G, u[i], storePreds);
-        }
-        DEBUG("running shortest path algorithm for node ", u[i]);
+    sampleNewPaths(0, r);
+    hasRun = true;
+}
 
-        sssp[i]->setTargetNode(v[i]);
-        sssp[i]->run();
-        if (sssp[i]->distances[v[i]] > 0) { // at least one path between {u, v} exists
-            DEBUG("updating estimate for path ", u[i], " <-> ", v[i]);
-            // random path sampling and estimation update
-            sampledPaths[i].clear();
-            node t = v[i];
-            while (t != u[i])  {
-                // sample z in P_u(t) with probability sigma_uz / sigma_us
-                std::vector<std::pair<node, double> > choices;
-                if (storePreds) {
-                    for (node z : sssp[i]->previous[t]) {
-                        // workaround for integer overflow in large graphs
-                        bigfloat tmp = sssp[i]->numberOfPaths(z) / sssp[i]->numberOfPaths(t);
-                        double weight;
-                        tmp.ToDouble(weight);
+void DynApproxBetweenness::update(GraphEvent e) {
+    std::vector<GraphEvent> batch(1);
+    batch[0] = e;
```

Comment:
Here you can just use:
```suggestion
    std::vector<GraphEvent> batch = {e};
```